security/src/main/java/io/micronaut/security/audit/UserAuditingEntityEventListener.java

```diff
@@ -83,23 +87,41 @@ public boolean preUpdate(@NonNull EntityEventContext<Object> context) {
         };
     }
 
-    private void autoPopulateUserIdentity(@NonNull EntityEventContext<Object> context, boolean isUpdate) {
-        if (securityService.isAuthenticated()) {
-            final RuntimePersistentProperty<Object>[] applicableProperties = getApplicableProperties(context.getPersistentEntity());
-            for (RuntimePersistentProperty<Object> persistentProperty : applicableProperties) {
-                if (isUpdate) {
-                    if (!persistentProperty.getAnnotationMetadata().booleanValue(AutoPopulated.class, AutoPopulated.UPDATEABLE).orElse(true)) {
-                        continue;
+    private void populate(@NonNull EntityEventContext<Object> context,
+                          @NonNull Class<? extends Annotation> listenerAnnotation) {
+        securityService.getAuthentication().ifPresent(authentication -> {
+            Map<Class<?>, Object> valueForType = new HashMap<>();
+            for (RuntimePersistentProperty<Object> persistentProperty : getApplicableProperties(context.getPersistentEntity())) {
+                if (shouldSetProperty(persistentProperty, listenerAnnotation)) {
+                    final BeanProperty<Object, Object> beanProperty = persistentProperty.getProperty();
+                    Object value = valueForType.computeIfAbsent(beanProperty.getType(), type -> convert(authentication, beanProperty));
+                    if (value != null) {
+                        context.setProperty(beanProperty, value);
                     }
                 }
+            }
+        });
+    }
 
-                final BeanProperty<Object, Object> beanProperty = persistentProperty.getProperty();
-                getCurrentUserIdentityForProperty(beanProperty).ifPresent(identity -> context.setProperty(beanProperty, identity));
+    @Nullable
+    private Object convert(@NonNull Authentication authentication, @NonNull BeanProperty<Object, Object> beanProperty) {
+        try {
+            return conversionService.convertRequired(authentication, beanProperty.getType());
+        } catch (ConversionErrorException e) {
```

Comment:
The next question is, should we actually propagate this error up to `preUpdate` and `prePersist` and then return `false` from them so that the data operation does not proceed?

Reply:
changed it here: https://github.com/micronaut-projects/micronaut-security/pull/1539/commits/cfa271e10b8e84b8cb49040c6587fe4053fc0e8b